File: test/address-cell-storage-oom.f
\ A real OS allocation refusal, after successful growth to the inline boundary.
\ Python supplies only Linux prlimit; the allocation under test is the engine's.
require lib/test.f
require lib/engine-candidate.f
package ADDRESS-CELL-STORAGE-OOM
create OUT 1024 allot
create ERR 4096 allot
: RUN ( -- )
   T-RESET
   HB-TARGET-LINUX? if
      PROC-ARGV-RESET
      s" test/address-cell-storage-oom.py" >LEN PROC-ARGV+
      ENGINE-CANDIDATE:PATH$ >LEN PROC-ARGV+
      s" /usr/bin/python3" >LEN OUT 1024 >LEN ERR 4096 >LEN 60000 >MS
      RUN-ARGV-CAPTURE
      MATCH result
         ok OF PCAP-CAPTURED:UNMAKE {: outu:len erru:len :}
            erru LEN>N 0 T=
            OUT outu LEN>N S\" address-cell-oom: ok\n" T$=
         ENDOF
         err OF PCAP-FAILED:UNMAKE {: outu:len erru:len rc:rc :}
            OUT outu LEN>N type ERR erru LEN>N type
            rc RC>N 0 T=
         ENDOF
      ;MATCH
   then
   T-REPORT ;
RUN
;package

File: test/address-cell-storage-oom.py
"""Linux-only OS boundary: deny the child's next mapping after its ready marker."""
import pathlib
import resource
import selectors
import subprocess
import sys

child_source = pathlib.Path(__file__).with_name("address-cell-oom-child.f")
child = subprocess.Popen([sys.argv[1], "--load", str(child_source)],
                         stdin=subprocess.PIPE, stdout=subprocess.PIPE,
                         stderr=subprocess.PIPE, text=True)
try:
    with selectors.DefaultSelector() as ready:
        ready.register(child.stdout, selectors.EVENT_READ)
        assert ready.select(30), "child did not reach the allocation boundary"
    assert child.stdout.readline() == "address-cell-oom: ready\n"
    status = pathlib.Path(f"/proc/{child.pid}/status").read_text()
    size_kib = next(int(row.split()[1]) for row in status.splitlines()
                    if row.startswith("VmSize:"))
    resource.prlimit(child.pid, resource.RLIMIT_AS, (size_kib * 1024,) * 2)
    out, err = child.communicate("x", timeout=30)
    assert child.returncode == 96, (child.returncode, out, err)
    assert out == "", repr(out)
    assert err == "hb: address-cell storage allocation failed\n", repr(err)
finally:
    if child.poll() is None:
        child.kill()
    child.wait()
print("address-cell-oom: ok")
